### PR TITLE
Fix a couple of CommentAsPatch issues. 

### LINKED ISSUE #14711: PL/SQL double dash comments in diff
Hi.
I use Gitea 1.13.2 on top of PostgreSQL.
A comment on a pull request including an Oracle pl/sql file with "--" comments results in something like this in DB in comment.patch:

```
diff --git a/sql/FUND/PACKAGES/CORRECTIONS.pkb b/sql/FUND/PACKAGES/CORRECTIONS.pkb
--- a/sql/FUND/PACKAGES/CORRECTIONS.pkb
+++ b/sql/FUND/PACKAGES/CORRECTIONS.pkb
---  some comment
---  some other comment
@@ -2061,839 +2119,733 @@
     END LOOP;
 
     IF gv_test_type THEN
-      Dr.begin ('Debit correction summary');
+      dr.begin ('Debit correction summary');
@@ -2703,0 +2522,4 @@
+                            l_register,
+                            v_hssr,
+                            vt_events (i_zd).oper_type,
+                            correction_cost.GC_ALG_PERC_SUM_BR_M, -- TODO
```

What happens next is that presumably the pulls panel treats the two "some comment" lines like the file locations above it and tries to link to them. The effect:

`template: repo/issue/view_content/comments:470:20: executing "repo/issue/view_content/comments" at <CommentMustAsDiff (index $comms 0)>: error calling CommentMustAsDiff: runtime error: invalid memory address or nil pointer dereference`

Manual removal of the two lines results in the pull request displayed properly.
Please check.

### LINKED ISSUE #14812: Commenting on a diff of a file with spaces separating short tokens makes the PR page inaccessible
Rendering the PR page errors with `template: repo/issue/view_content/comments:470:20: executing "repo/issue/view_content/comments" at <CommentMustAsDiff (index $comms 0)>: error calling CommentMustAsDiff: runtime error: slice bounds out of range [2:1]`

I was unable to get a proper stacktrace (or logs), but I manually traced the offending slice indexing to [this line](https://github.com/go-gitea/gitea/blob/014313134f72ebb2ca57eb2e5cd42726231f8fff/services/gitdiff/gitdiff.go#L867).

The git-diff output filename extractor (`readFileName`)  cannot properly split the old name from the new name (and panics because instead of finding `b/<filename>` to remove `b/` it finds the single letter bit of the old filename).

git-diff outputs **unquoted** paths because Gitea [sets `core.quotePath` (git config) to `false`](https://github.com/go-gitea/gitea/blob/50208e903ae9aa931db3d93bde7032e766687cae/modules/git/git.go#L150). Interestingly, `readFileName` has code handling quoted paths already. It did work too when I changed the offending patch in the comments table to have quoted paths.

## Steps to reproduce:
1. Open a PR creating a file whose name is something like  `sdafdsaf z` `1 - the first document.txt` (anything in the format`<any amount of characters that are not a space> <1 non-space character>[ <anything>]` works as far as I can tell)
2. Comment on any line of this new file in the PR
3. The conversation tab of this PR now gives a 500 error when accessed

## Fix suggestion
Maybe there's a really good reason to keep it, but imho leaving all paths returned by git quoted (`core.quotePath=true`) is the best option for sane path handling everywhere. 

It **may** already be ok to do so: setting it to true didn't break any tests on my machine at least.

- Gitea version (or commit ref): 1.13.2
- Git version: 2.26.2
- Operating system: Gentoo
    * using official ebuild with USE="acct build-client filecaps pam -sqlite"
    * started by OpenRC
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [x] SQLite (also tested)
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes [example](https://try.gitea.io/templatepanic/templatepanic/pulls/3)
  - [ ] No
- Log gist: no logs appear (not even with Trace everything)

### PR DESCRIPTION
**Pull request by zeripath**
## CutDiffAroundLine makes the incorrect assumption that `---` and `+++` always represent part of the header of a diff. 

This PR adds a flag to its parsing to prevent this problem and adds a streaming parsing technique to CutDiffAroundLine using an io.pipe instead of just sending data to an unbounded buffer.

## Handle unquoted comment patch filenames
    
When making comment patches unfortunately the patch does not always quote the filename. This makes the diff --git header ambiguous again.
    
This PR finally adds handling for ambiguity in to parse patch
    
Fix #14711
Fix #14812
    
Signed-off-by: Andrew Thornton <art27@cantab.net>

-------------------------

### CODE REVIEW

**Comment by zeripath** [comment]
Cutdiffaroundline also needs to have some protection against truly pathological diffs. E.g. single line minimised js file becomes unminimised and you comment on one of the lines

It's also essentially a duplicating some of the parsepatch code.

File: modules/git/diff_test.go [hunk 1]
```diff
@@ -37,18 +56,44 @@ func TestCutDiffAroundLine(t *testing.T) {
...
+	minusDiff, _ = CutDiffAroundLine(strings.NewReader(breakingDiff), 3, false, 4)
+
+	expected = `diff --git a/aaa.sql b/aaa.sql
+--- a/aaa.sql
++++ b/aaa.sql
+@@ -1,9 +1,10 @@
+ --some comment
+--- some comment 5
++--some coment 2
++--some comment 3`
```
Comment (zeripath) [inline modules/git/diff_test.go]:
```suggestion
+-- some comment 3`
```

File: modules/git/diff.go [hunk 2]
```diff
@@ -125,30 +125,40 @@ var hunkRegex = regexp.MustCompile(`^@@ -(?P<beginOld>[0-9]+)(,(?P<endOld>[0-9]+
 
 const cmdDiffHead = "diff --git "
 
-func isHeader(lof string) bool {
-	return strings.HasPrefix(lof, cmdDiffHead) || strings.HasPrefix(lof, "---") || strings.HasPrefix(lof, "+++")
+func isHeader(lof string, inHunk bool) bool {
+
```
Comment (lunny) [inline modules/git/diff.go:129]:
```suggestion
---
```

Comment (6543) [inline modules/git/diff.go]:
```suggestion
```

**Review by 6543 — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:

**Pull request merged**